src/Node/SetNode.php

```diff
@@ -33,9 +33,15 @@ public function __construct(bool $capture, Node $names, Node $values, int $linen
         $safe = false;
         if ($capture) {
             $safe = true;
-            if ($values instanceof TextNode) {
+            if (Node::class === get_class($values) && !count($values)) {
```

Comment:
See this makes me think we should really investigate having a dedicated node class for representing a statement list in 4.x (as suggested in https://github.com/twigphp/Twig/pull/4292#discussion_r1751616965), as we have this special case of `Node::class === get_class($values)` in more and more places